cas-compute/src/symbolic/expr/mod.rs

```diff
@@ -436,6 +441,50 @@ impl PartialEq for Expr {
     }
 }
 
+/// Implements the [`std::hash::Hash`] trait for the [`Expr`] enum, which represents
+/// different types of mathematical expressions. This implementation is necessary
+/// because [`PartialEq`] is also manually implemented, and the automatically
+/// derived [`Hash`] implementation would not guarantee consistency with [`PartialEq`].
+impl std::hash::Hash for Expr {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        // Hash the discriminant to differentiate between enum variants.
+        let self_discr = std::mem::discriminant(self);
+        std::hash::Hash::hash(&self_discr, state);
+
+        // Hash the data associated with each variant.
+        match self {
+            Expr::Primary(val) => std::hash::Hash::hash(val, state),
+            Expr::Add(val) | Expr::Mul(val) => {
+                // For Add and Mul, the order of the elements should not affect the hash.
+                // Therefore, we'd have to sort the individual elemets hashes before
+                // hashing them.
+
+                let mut hashes: Vec<u64> = val
+                    .iter()
+                    .map(|expr| {
+                        let mut hasher = DefaultHasher::new();
+
+                        std::hash::Hash::hash(expr, &mut hasher);
+
+                        hasher.finish()
+                    })
+                    .collect();
+
+                // To satisfy order invariance.
+                hashes.sort_unstable();
+
+                for hash in hashes {
+                    std::hash::Hash::hash(&hash, state);
+                }
+            }
+            Expr::Exp(val_base, val_exp) => {
+                std::hash::Hash::hash(val_base, state);
+                std::hash::Hash::hash(val_exp, state);
+            }
+        }
+    }
+}
+
```

Comment:
This strategy works well, however it does end up taking `O(n log n)` time and can end up allocating memory for the hashes, which isn't the most ideal for a hash function. I am somewhat inclined to accept this, but if we can modify this approach to take `O(n)` time while also preventing it from allocating memory, that would be splendid.

Reply:
Yeah, `O(n log n)` ain't bad but it can be a lot better. That complexity is due to multiple nested sorts. So, I'd look into that approach you shared.

Reply:
I am not feeling satisfied with the "just XOR the hashes" approach. Might have to look a bit into [this](https://stackoverflow.com/a/7448843/18016046) and more.

Reply:
@ElectrifyPro Please take a look